src/ucp/stream/stream_recv.c

```diff
@@ -0,0 +1,133 @@
+/**
+ * Copyright (C) Mellanox Technologies Ltd. 2017.  ALL RIGHTS RESERVED.
+ *
+ * See file LICENSE for terms.
+ */
+
+#include <ucp/core/ucp_ep.h>
+#include <ucp/core/ucp_worker.h>
+#include <ucp/core/ucp_context.h>
+#include <ucp/core/ucp_request.h>
+
+#include <ucs/datastruct/mpool.inl>
+#include <ucs/debug/profile.h>
+
+#include "../tag/eager.h"
+
+
+typedef struct {
+    uint64_t    uuid;
+} UCS_S_PACKED ucp_stream_eager_hdr_t;
+
+UCS_PROFILE_FUNC(ucs_status_ptr_t, ucp_stream_recv_data_nb,
+                 (ep, length), ucp_ep_h ep, size_t *length)
+{
+    ucp_recv_desc_t  *rdesc;
+    ucs_status_ptr_t ret;
+
+    UCP_THREAD_CS_ENTER_CONDITIONAL(&ep->worker->mt_lock);
+
+    if (ucs_list_is_empty(&ep->stream_data)) {
+        ret = UCS_STATUS_PTR(UCS_OK);
+        goto out;
+    }
+
+    rdesc = ucs_list_next(&ep->stream_data, ucp_recv_desc_t,
+                          list[UCP_RDESC_HASH_LIST]);
+    ucs_list_del(&rdesc->list[UCP_RDESC_HASH_LIST]);
+    *length = rdesc->length;
+    ret = (ucs_status_ptr_t)((uintptr_t)(rdesc + 1) + rdesc->hdr_len);
+
+out:
+    UCP_THREAD_CS_EXIT_CONDITIONAL(&ep->worker->mt_lock);
+
+    return ret;
+}
+
+UCS_PROFILE_FUNC_VOID(ucp_stream_data_release, (ep, data),
+                      ucp_ep_h ep, void *data)
+{
+    /* TODO: make const header offset */
+    ucp_recv_desc_t *rdesc;
+    rdesc = (ucp_recv_desc_t *)((uintptr_t)(data -
+                                            sizeof(ucp_stream_eager_hdr_t))) - 1;
+
+    UCP_THREAD_CS_ENTER_CONDITIONAL(&ep->worker->mt_lock);
+
+    if (ucs_unlikely(rdesc->flags & UCP_RECV_DESC_FLAG_UCT_DESC)) {
+        uct_iface_release_desc((char*)rdesc - sizeof(ucp_eager_sync_hdr_t));
+    } else {
+        ucs_mpool_put_inline(rdesc);
+    }
+
+    UCP_THREAD_CS_EXIT_CONDITIONAL(&ep->worker->mt_lock);
+}
+
+static UCS_F_ALWAYS_INLINE ucs_status_t
+ucp_eager_handler(void *arg, void *data, size_t length, unsigned am_flags,
+                  uint16_t flags, uint16_t hdr_len)
+{
+    ucp_worker_h            worker  = arg;
+    ucp_ep_h                ep;
+    ucp_stream_eager_hdr_t  *eager_hdr;
+    ucp_recv_desc_t         *rdesc;
+    size_t                  recv_len;
+    khiter_t                hash_it;
+
+    ucs_status_t status = UCS_ERR_NOT_IMPLEMENTED;
+
+    UCP_THREAD_CS_ENTER_CONDITIONAL(&worker->mt_lock);
+
+    ucs_assert(length >= hdr_len);
+    recv_len = length - hdr_len;
+
+    if (ucs_unlikely(am_flags & UCT_CB_PARAM_FLAG_DESC)) {
```

Comment:
yes, but here 2 reason why i did not do that:
1. it will affect TAG code make this PR bigger and harder for review
2. need to add extra call on fast path or new one common.inl file, it also complicates review.

i'll add "todo" and fix in next PRs with refactoring